### PR TITLE
Add GKeyFile to the API

### PR DESCRIPTION
**Pull request by hfiguiere**

### CODE REVIEW

**Comment by hfiguiere** [comment]
This should fix issue #135. See the discussion over there.

**Comment by hfiguiere** [comment]
I currently use a non HEAD/master commit for gir as there is a PR gtk-rs/gir#265

**Comment by hfiguiere** [comment]
This is causing the failure on TravisCI. When that PR lands, I'll change the gitsubmodule commit to address this.

**Comment by GuillaumeGomez** [comment]
All this Makefile and submodule things aren't very encouraging. Also, we won't merge anything as long as CIs don't say 'ok'.

File: src/source.rs [hunk 1]
```diff
@@ -6,7 +6,8 @@ use std::cell::RefCell;
 use std::mem::transmute;
 use std::process;
 use std::thread;
-use glib_ffi::{self, gboolean, gpointer};
+use ffi as glib_ffi;
+use ffi::{self, gboolean, gpointer};
```
Comment (EPashkin) [inline src/source.rs]:
```
warning: unused import, #[warn(unused_imports)] on by default
  --> src\source.rs:10:11
   |
10 | use ffi::{self, gboolean, gpointer};
   |           ^^^^
```

File: Gir.toml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+]
+
+[[object]]
+name = "GLib.KeyFile"
+status = "generate"
+       [[object.function]]
+       name = "load_from_data_dirs"
+       ignore = true
+       [[object.function]]
+       name = "load_from_dirs"
```
Comment (EPashkin) [inline Gir.toml]: Its not in manual part?
Compiling example looks like this:
```rust
   pub fn load_from_dirs<T: AsRef<std::path::Path>, U: AsRef<std::path::Path>>(&self, file: T, search_dirs: &[U], flags: KeyFileFlags) -> Result<path::PathBuf, Error> {
        unsafe {
            let search_dirs: Vec<&std::path::Path> = search_dirs.iter().map(|p| p.as_ref()).collect();
            let mut error = ptr::null_mut();
            let mut full_path: *mut libc::c_char = ptr::null_mut();
            let _ = ffi::g_key_file_load_from_dirs(self.to_glib_none().0,
                                                        file.as_ref().to_glib_none().0,
                                                        search_dirs.to_glib_none().0,
                                                        &mut full_path,
                                                        flags.to_glib(), &mut error);
            if error.is_null() {
                let path: String = from_glib_full(full_path);
                Ok(path::PathBuf::from(path))
            } else {
                Err(from_glib_full(error))
            }
        }
   }
```

File: src/key_file.rs [hunk 3]
```diff
@@ -0,0 +1,38 @@
+
```
Comment (EPashkin) [inline src/key_file.rs]: Add copyright lines, please

**Comment by hfiguiere** [comment]
As I said, the CI will pass for you merge gtk-rs/gir#265, which is fine to wait on.

File: Gir.toml (hunk 2 above)
Comment (hfiguiere) [inline Gir.toml]: I was meaning to remove it from the Gir.toml, but then I'll do quite the opposite. Will update the PR

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,17 @@
+GIR = gir/target/bin/gir
+GIR_SRC = gir/Cargo.toml gir/Cargo.lock gir/build.rs $(shell find gir/src -name '*.rs')
+GIR_FILES = gir-files/Gtk-3.0.gir
```
Comment (gkoz) [inline Makefile]: This would be `gir-files/GLib-2.0.gir`.

File: Gir.toml [hunk 5]
```diff
@@ -0,0 +1,35 @@
+[options]
+girs_dir = "gir-files"
+library = "GLib"
+version = "2.0"
+min_cfg_version = "2.32"
+target_path = "."
+work_mode = "normal"
+
+generate = [
+    "GLib.KeyFileFlags",
```
Comment (gkoz) [inline Gir.toml]: I think `GLib.KeyFileError` ought to be here as well.

File: Gir.toml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+status = "generate"
+       [[object.function]]
+       name = "load_from_data_dirs"
+       ignore = true
+       [[object.function]]
+       name = "load_from_dirs"
+       ignore = true
+       [[object.function]]
+       name = "error_quark"
+       ignore = true
```
Comment (gkoz) [inline Gir.toml]: As long as the error enum is generated this bit should not be necessary.

File: Makefile (hunk 4 above)
Comment (hfiguiere) [inline Makefile]: doh. ! Will fix too

**Comment by EPashkin** [comment]
Gir part merged

File: Gir.toml (hunk 6 above)
Comment (gkoz) [inline Gir.toml]: Oh, is the function filed under `GKeyFile` instead of `GKeyFileError`? This might need some workaround in gir.

Comment (hfiguiere) [inline Gir.toml]: generated binding return a "Quark" whose import fail unless I reexport it from ffi::GQuark, but @EPashkin told me not to. So I did it manually.... To me is is the generator that should be fixed...

File: Gir.toml (hunk 2 above)
Comment (hfiguiere) [inline Gir.toml]: This is now in the PR

File: src/key_file.rs (hunk 3 above)
Comment (hfiguiere) [inline src/key_file.rs]: done too

File: Gir.toml (hunk 5 above)
Comment (hfiguiere) [inline Gir.toml]: Good catch, I added this one.

File: Gir.toml (hunk 6 above)
Comment (gkoz) [inline Gir.toml]: Since the error enum is generated correctly never mind, this `ignore` seems the least hairy solution.

**Comment by hfiguiere** [comment]
My bad I'll fix the CI.

**Comment by hfiguiere** [comment]
it is actually a codegen bug :-/

**Comment by hfiguiere** [comment]
It need this change in gir
````
diff --git a/src/codegen/enums.rs b/src/codegen/enums.rs
index 2ef5175..efdad75 100644
--- a/src/codegen/enums.rs
+++ b/src/codegen/enums.rs
@@ -134,10 +134,16 @@ impl FromGlib<ffi::{ffi_name}> for {name} {{
     if let Some(ref get_quark) = enum_.error_domain {
         let get_quark = get_quark.replace("-", "_");
         let has_failed_member = members.iter().any(|m| m.name == "Failed");
+        let glib_sys_module: &str;
+        if env.namespaces.glib_ns_id == namespaces::MAIN {
+            glib_sys_module = "ffi";
+        } else {
+            glib_sys_module = "glib_ffi";
+        }
 
         try!(version_condition(w, env, enum_.version, false, 0));
         try!(writeln!(w, "impl ErrorDomain for {name} {{
-    fn domain() -> glib_ffi::GQuark {{
+    fn domain() -> {glib_sys_module}::GQuark {{
         {assert}unsafe {{ ffi::{get_quark}() }}
     }}
 
@@ -146,7 +152,8 @@ impl FromGlib<ffi::{ffi_name}> for {name} {{
     }}
 
     fn from(code: i32) -> Option<Self> {{
-        {assert}match code {{", name = enum_.name, get_quark = get_quark, assert = assert));
+        {assert}match code {{", name = enum_.name, get_quark = get_quark, assert = assert,
+                      glib_sys_module = glib_sys_module ));
 
         for member in &members {
             try!(version_condition(w, env, member.version, false, 3));

````

So here is what I propose:

- I remove ````GLib.KeyFileError```` from the generate Gir.toml
- we get this merge
- I create a new for the above, and one for gir that will be needed to solve the above issue.

**Comment by EPashkin** [comment]
Agree with merge without `GLib.KeyFileError`.
`Gir` can also fixed in enum.rs. Not sure what better.
```diff
diff --git a/src/codegen/enums.rs b/src/codegen/enums.rs
index 2ef5175..e0c0f1c 100644
--- a/src/codegen/enums.rs
+++ b/src/codegen/enums.rs
@@ -19,7 +19,8 @@ pub fn generate(env: &Env, root_path: &Path, mod_rs: &mut Vec<String>) {
         try!(writeln!(w, ""));
         try!(writeln!(w, "use ffi;"));
         if env.namespaces.glib_ns_id == namespaces::MAIN {
-            try!(writeln!(w, "use error::ErrorDomain;
+            try!(writeln!(w, "use ffi as glib_ffi;
+use error::ErrorDomain;
 use translate::*;"));
         } else {
             try!(writeln!(w, "use glib_ffi;
```

File: Gir.toml (hunk 5 above)
Comment (hfiguiere) [inline Gir.toml]: Actually see the comments below. It needs another fix in gir.

**Comment by gkoz** [comment]
Well, I vote for the smaller diff ;)

**Comment by hfiguiere** [comment]
Ok, I'll check that.

**Comment by hfiguiere** [comment]
ok this is the updated PR

File: src/key_file.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+use KeyFile;
+
+impl KeyFile {
+
+    pub fn error_quark() -> ffi::GQuark {
+        unsafe {
+            ffi::g_key_file_error_quark()
+        }
+    }
```
Comment (gkoz) [inline src/key_file.rs]: Oh, I guess we've overlooked this bit. The function doesn't need to be exposed, the generated enum whenever it gets added covers that.

File: src/lib.rs [hunk 8]
```diff
@@ -130,9 +132,13 @@ pub mod shared;
 #[macro_use]
 pub mod object;
 
+pub use auto::*;
+mod auto;
+
 mod bytes;
 pub mod error;
 mod file_error;
+pub mod key_file;
```
Comment (gkoz) [inline src/lib.rs]: We keep the modules private unless they provide some extensions (implementation details) needed by other `gtk-rs` crates not reexported in `lib.rs`.

File: src/key_file.rs (hunk 7 above)
Comment (hfiguiere) [inline src/key_file.rs]: No. The problem is that gir generate code that doesn't build:
````
src/auto/key_file.rs:6:5: 6:10 error: unresolved import `Quark`. There is no `Quark` in the crate root [E0432]
src/auto/key_file.rs:6 use Quark;
                           ^~~~~
src/auto/key_file.rs:6:5: 6:10 help: run `rustc --explain E0432` to see a detailed explanation
````

Fixing gir should help getting rid of that manual wrapper.

File: src/lib.rs (hunk 8 above)
Comment (hfiguiere) [inline src/lib.rs]: this is fixed now.

**Comment by EPashkin** [comment]
:+1: for this state.
Complete remove `KeyFile::error_quark` can be done with adding `GLib.KeyFileError` and `KEY_FILE_DESKTOP_*`.
Or we can merge PR266 and do it in this PR.

**Comment by hfiguiere** [comment]
Updated PR with the constant.
Wrapped manually a function to take a path instead of a string.

File: Cargo.toml [hunk 9]
```diff
@@ -27,3 +28,7 @@ version = "0.3.1"
 [dependencies.gobject-sys]
 git = "https://github.com/gtk-rs/sys"
 version = "0.3.1"
+
+[features]
+v2_38 = [ "glib-sys/v2_38", "gio-sys/v2_38", "gobject-sys/v2_38" ]
+v2_40 = [ "v2_38", "glib-sys/v2_40", "gio-sys/v2_40" ]
```
Comment (EPashkin) [inline Cargo.toml]: Empty line at end, pls

File: Gir.toml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+       [[object.function]]
+       name = "save_to_file"
+       # file parameter needs to be a PathBuf
+       ignore = true
```
Comment (EPashkin) [inline Gir.toml]: Empty line, pls

**Comment by hfiguiere** [comment]
addressed the last two comments.

**Comment by EPashkin** [comment]
@hfiguiere, as @gkoz or @GuillaumeGomez still not merged this PR, can you add `GLib.KeyFileFlags` and remove manual `error_quark` here?

**Comment by hfiguiere** [comment]
no, because it is not solved by the current gir.... ie it doesn't build.

**Comment by EPashkin** [comment]
Strange, PR266 merged and all builds fine for me.

**Comment by gkoz** [comment]
I believe updating the gir submodule to the latest master fixes the `KeyFileError` issue.

**Comment by hfiguiere** [comment]
I tried that before commenting. If it has changed since, maybe. But 13 hours ago it didn't fix it.

**Comment by gkoz** [comment]
At least it's worked for me right now. But submodules can be a bit confusing sometimes.

**Comment by hfiguiere** [comment]
I did pull master, rebuild the generated filed. Still not fixing the problem. Problem still in Gir and I'm not familiar with it, so I haven't dug into the problem yet. I thinks perfect is the enemy of good in that case as we are still better than last time.

I have already pushed by code to use the AP this PR provides.

**Comment by gkoz** [comment]
Thanks for contributing! I'll sort out the minor things later then.
@GuillaumeGomez any objections?

**Comment by hfiguiere** [comment]
Closed by way of #137

**Pull request closed without merging**